[MRG+1] Added _fit_svd_solver variable to PCA

njkevlani opened the pull request:
#### Fixes #11223 Expose solver used with PCA(svd_solver='auto')

_fit_method will be set as soon as fit(X) method is called. Depending upon the value passes for svd_solver and input array X, value of _fit_method will be set.

njkevlani committed:
Added _fit_method variable

_fit_method will be set as soon as fit(X) method is called. Depending upon the value passes for svd_solver and input array X, value of _fit_method will be set.

qinhanmin2014 commented on sklearn/decomposition/pca.py:
```diff
@@ -388,7 +388,7 @@ def _fit(self, X):
 
-        # Handle svd_solver
+        # Handle svd_solver and _fit_method
         svd_solver = self.svd_solver
```
We can replace variable ``svd_solver`` with the new ``self._fit_method``

rth replied:
Everywhere? IMO `svd_solver` is more concise/readable..

njkevlani replied:
`svd_solver` is just being a local variable in `_fit()` method. If we replace it with a class variable (say `self._fit_svd_solver` for now), it will serve both the purposes, doing `svd_solver`'s current task and expose the solver used. 

Moreover, `NearestNeighbours` also handle `algorithm='auto'` in this manner.

rth replied:
Yes, but you would have to do 4-5 replacements of `svd_solver` with `self._fit_svd_solver` to save one assignment, `self._fit_svd_solver = svd_solver` at the end (current solution). In any case that's not too important, either way would be fine, up to you.

qinhanmin2014 commented:
Generally we need tests for a PR related to source code, but since ``_fit_method`` is private and we also don't have related test in ``sklearn.neighbor``, I'm fine not to add a test here.

jnothman commented:
I'm not sure that _fit_method is a great name... I'd rather have "solver in
there"

njkevlani commented:
@jnothman Will it be okay if we overwrite value of `self.svd_solver` if it is passed as `"auto"`? By doing this we will solve the issue and maintain less variables.

rth commented:
> Will it be okay if we overwrite value of self.svd_solver if it is passed as "auto"?

No that wont work, because, when running,
```py
pca = PCA(svd_solver='auto')
pca.fit(X1)
pca.fit(X2)
```
you wouldn't want the solver in the second fit to depend on the choice in the first one.
> I'm not sure that _fit_method is a great name... I'd rather have "solver in
there"

+1 maybe `_fit_svd_solver` ? Not great, but I can't think of something better.

> Generally we need tests for a PR related to source code, but since _fit_method is private and we also don't have related test in sklearn.neighbor, I'm fine not to add a test here.

Yes, though a test for two repeated fits (see above) behave consistently for the solver wouldn't hurt. Unless there is already a test like that.

njkevlani commented:
@rth 
> No that wont work, because, when running,

>`pca = PCA(svd_solver='auto')`
>`pca.fit(X1)`
>`pca.fit(X2)`

Agree with that. How about [this](https://github.com/scikit-learn/scikit-learn/pull/11225#discussion_r194251389)?

qinhanmin2014 commented:
> maybe _fit_svd_solver ? 

I'm OK with it and actually don't care much about the name (since it's private). Maybe we can also consider ``self._svd_solver``? So we have ``self.svd_solver`` for parameter ``svd_solver`` and ``self._svd_solver`` for the solver used by PCA.

njkevlani committed:
Used _fit_svd_solver instade of svd_solver

Replacing svd_solver with self._fit_svd_solver let us keep track of svd_solver method even if it is passed as 'auto'

qinhanmin2014 commented on sklearn/decomposition/pca.py:
```diff
@@ -388,26 +388,26 @@ def _fit(self, X):
         else:
             n_components = self.n_components
 
-        # Handle svd_solver
```
Not a serious problem but I think you might leave the comment as it is. (``Handle _fit_svd_solver`` seems awkward from my side.)

qinhanmin2014 reviewed (approved):
LGTM, thanks @njkevlani

njkevlani committed:
Corrected comment

qinhanmin2014 commented on sklearn/decomposition/pca.py:
```diff
@@ -388,26 +388,26 @@ def _fit(self, X):
-        # Handle svd_solver
-        svd_solver = self.svd_solver
-        if svd_solver == 'auto':
+        # Handle _svd_solver
```
``_svd_solver`` -> ``svd_solver`` :)

njkevlani replied:
@qinhanmin2014 Fixed typo.

njkevlani committed:
fixed typo

njkevlani commented:
@qinhanmin2014 @rth Would you guys want me to change anything further?

rth reviewed (approved):
LGTM, merging. Thank you @njkevlani !
 
(I would have preferred to the keep the local `svd_solver` variable as less verbose, but since there is no consensus on that, this is fine.)
